File: 94-BinaryTreeInorderTraversal/94-BinaryTreeInorderTraversal.java
// Last updated: 8/13/2025, 8:23:00 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> list = new ArrayList();
        traversal(root,list);
        return list;
    }

    public void traversal(TreeNode root, ArrayList list){
        if(root==null){
            return;
        }
        traversal(root.left,list);
        list.add(root.val);
        traversal(root.right,list);
    }
}

File: 94-BinaryTreeInorderTraversal/94-BinaryTreeInorderTraversal.py
# Last updated: 8/13/2025, 8:22:54 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        curr = root
        result = []

        while curr is not None : 
            if curr.left is None : 
                result.append(curr.val)
                curr = curr.right
            
            else : 
                currp1 = curr.left
                while currp1.right is not None and currp1.right!= curr : 
                    currp1 = currp1.right
                
                if currp1.right is None : 
                    currp1.right = curr
                    
                    curr = curr.left
                
                else : 
                    currp1.right = None
                    result.append(curr.val)
                    curr = curr.right
        
        return result